### PR TITLE
Modernize code base by using new features of the Factorio engine

### LINKED ISSUE #39: Code base is old and doesn't make use of new features
This leads to two problems:

The first one is that it lacks some of the new, very useful items that have been added since the majority of the mod was written. The loader 1x1, infinity chests, infinity pipes, linked chests and belts, etc.

The second one is related to the first: since it doesn't make use of any of those new features, a lot of stuff in the mod is handled by scripting, leading to large performance issues from the mod. Enough that a lot of people have switched to other, similar mods. Most of the items can be changed to make use of base game prototypes with minimal scripting and little or no change to functionality.

For instance, one very easy change that reduces scripting load significantly is to replace the current system of burner energy types with scripted refueling to instead use the 'void' energy type, which requires no powering. It likely didn't exist when the prototypes were written, and can be swapped in with no changes to the user experience.

Infinity chests can replace several of the chest types with no/minimal need for scripting. Fluid voids can be replaced by infinity pipes. Heat sources/voids can be replaced by Heat Interfaces. All with very significant savings since the complicated bits are managed in-engine rather than via LUA.

The entities aren't free processing-wise, but they're significantly less UPS intensive and more in line with other mods.

### PR DESCRIPTION
**Pull request by Silari**
Resolves #39 

The pull includes everything that should be required to release it as 1.8.1, including info.json and changelog.txt changes.

UPS test results - measuring time spent in creative-mod on an otherwise empty map:
Empty map - 0.065ms average with 1.8.0, to 0.06ms average with 1.8.1
Map with 400: fluid source into tank into fluid void - ~3.5ms average, to ~0.07ms avg
Map with 805 super radars - ~1.9ms average, to ~0.07ms average
Map with 350: inf chest into loader onto blue belt into loader into void chest - ~0.13ms avg, to ~0.068ms
Map with 140 heat exchangers + pipes + heat pipes + 14 heat source + 14 heat void + 16 vanilla infinity pipe, making 347K steam per minute: 0.17ms avg, to 0.078ms and still making 347K steam.

Full change log is in the pull, but here's the truncated change log:
```
  Features:
    - Added a recipe for the vanilla infinity chest, as well as requester and provider variants, a version of the Loader1x1, a version of the linked chest, and a version of linked belts. 
  Optimizations:
    - Removed all scripting for fluid-source, fluid-void, super-radar, super-beacon, super-roboport, energy-void, energy-source, passive-energy-void, passive-energy-source. They no longer require it. See details below.
    - Changed all creative entities requiring power to use a void energy source.
    - Changed Heat Source and Heat Void to use the Heat Interface prototype instead. 
        - These look different now due to the differences in how heat-interfaces work - they don't have separate connection sprites like heat pipes do, so they are locked into the 4 way sprite.
    - Changed creative-chest, creative-provider-chest, void chest, void storage chest, and void requester chest to use the vanilla infinity chest. No changes in usage should be present.
        - Removed all on-tick scripting for these entities. There is a quick script ran on placement that sets them up, and GUI scripting for the creative chests.
        - creative-chest and creative-provider-chest have two minor regressions: 
            - Copy/paste no longer works since the base game considers them non-operable. As it appears this only set the group number, and not the filters, this is fairly minor.
            - The inventory display mode also doesn't do anything for them, since infinity chests sort their contents. Technically it DOES have a slight change in how the filters are set, but as those are only viewable in the editor this isn't particularly useful.
    - Changed Fluid Void to use the vanilla infinity pipe.
        - Old fluid voids will NOT be automatically swapped to the new one - due to changing to a pipe, new fluid voids will connect in all four directions, which may cause issues with previous builds. They will remain scripted and thus use significantly more UPS until replaced. Any newly built fluid voids will be the new style and not use scripting.
        - These are now a purple tinted pipe instead of the old graphics, due to the prototype change and requirement to connect in all four directions.
    - Energy void no longer produces an item. It should run constantly as long as it has power.
    - Active and Passive Energy Sources and Passive Energy Void are now electric interfaces and require no LUA scripting to function.
        - They no longer have a circuit connection due to this change.
```

### CODE REVIEW

File: scripts/creative-chest-util.lua [hunk 1]
```diff
@@ -310,6 +391,10 @@ function creative_chest_util.on_entity_copied_pasted(source_chest, destination_c
 
 	local source_chest_data, source_group_number =
 		creative_chest_util.get_creative_chest_data_group_number(source_chest, source_groups)
+    log(source_groups)
+    log(destination_groups)
+    log(source_chest_data)
+    log(source_group_number)
```
Comment (jodli) [inline scripts/creative-chest-util.lua:397]: Do we need these log calls?

File: scripts/gui-entity.lua [hunk 2]
```diff
@@ -252,7 +252,8 @@ local function add_creative_chest_group_number_and_update_gui(creative_chest_ent
 				chest_groups,
 				group_number,
 				index_in_group,
-				new_group_number
+				new_group_number,
+                creative_chest_entity
```
Comment (jodli) [inline scripts/gui-entity.lua:256]: Formatting?

**Review by jodli — COMMENTED** [review]
First of all: Thank you for your awesome contribution. I appreciate it very much!

Do you know of any other feature that was added to the base version of the mod, which is still implemented in a scripting way? That would make a new major version feel more like a milestone.

File: scripts/creative-chest-util.lua (hunk 1 above)
Comment (Silari) [reply scripts/creative-chest-util.lua]: Nope, was the last thing I did so I missed them. They're gone.

File: scripts/gui-entity.lua (hunk 2 above)
Comment (Silari) [reply scripts/gui-entity.lua]: Tabs vs spaces issue, changed to tabs to match the rest.

**Comment by Silari** [comment]
There wasn't a lot left that I could see that could be swapped without making major changes, which I wanted to avoid, or wasn't already handled with the stuff already available. 

Biggest one would be adding a recipe for the vanilla infinity pipe, same as there's one for the infinity chest. They have an advantage of being able to select any fluid but also setting the temperature to any allowed value. The new void pipes are one internally, but modified with a bigger fluid box so they remove fluid just as fast as the old ones. Fluid sources are still assemblers because those push out fluid faster - I believe because a) they have a ton of fluid inside due to the recipe size, and b) they have a base_level of 1 compared to the default 0 of pipes. Could be fixed maybe with proper fluid box settings on an infinity pipe, but tbh I didn't think about it until just now. 

Duplicating chests could have been changed to also use infinity chests - when placed throw them into a list like they are now, except when an item is put inside (in any slot) and the on_tick catches it, set the infinity filter to match and remove it from the list. That way there's only scripting required before the item is set. Downside is you can't unset it easily - could probably have repurposed the GUI checkbox for that easily though. Didn't think it was worth the time since you can just use an infinity chest to duplicate whatever.

Similarly the autofill requesters could have been changed to use infinity chests, but again we've got an infinity requester chest already. The autofill one still has a use in that it can set it's requests based on the circuit network, then have those requests filled by the script. Technically, that functionality could be moved to the infinity requester chest, but that would require it have on_tick scripting which is what this entire update is trying to avoid.

Matter sources and voids are still entirely scripted, but most uses can be replaced with an infinity chest and a loader easily - slightly larger but much more efficient. Can't replace a source/void pointing at a building/container to add/remove multiple items - mostly that's just useful to fill the rocket silo though.

The super boiler/cooler can't really be replaced with anything well. Infinity pipes can alter the temperature of whatever passes through, but unfortunately it REQUIRES it to be set to 'at least' mode with at least 1%, which of course would create free fluid. They COULD possibly be changed to furnaces so that whatever fluid is put in, it heats/cools to it max/default but that has some tradeoffs. Big one being that recipes work in set amounts, so if the recipe takes 100/gives 100 that would limit the input/output to 100 per tick. Setting it higher would increase the possible throughput, but it also means that it doesn't do anything until you get that much fluid.

No idea on the wands and alien attracters. There were some changes to selection tools and such which might be useful but nothing I can think of offhand.

Super personal roboport can't be changed to use a void energy source because that's apparently not valid for equipment.


That turned out way longer than expected.

**Comment by jodli** [comment]
> That turned out way longer than expected.

You're crazy! 🤯
Will try this out tomorrow and then I'll create a new package. Thank you so much! 🏅

> There wasn't a lot left that I could see that could be swapped without making major changes, which I wanted to avoid, or wasn't already handled with the stuff already available.

You seem to be somewhat interested in the future of this mod. Can talk about this offline, maybe on Discord (`jodli#7163`)?

**Comment by Silari** [comment]
Unfortunately between two jobs and transitioning I don't really have a lot of time/enthusiasm for modding for a while, not even my own stuff. That was why I didn't add code to automatically swap the old fluid voids to the new ones by checking for extra connections to ensure it was safe - not enough time to do much more than what I already had done.

And really this all came out of being annoyed that using a fluid source on an SE spaceship broke when it teleported since the fuel filtering broke.

Which is probably something that should be fixed for the remaining scripted entities by listening for the events, if possible.

**Pull request merged**